### DISCUSSION TITLE
Low Write and Read Performance with OrangeFS-2.9.8

### DISCUSSION THREAD

**Discussion by Jye-525** [waltligon/orangefs#96] [Q&A]
Hi,
I am setting up OrangeFS on a single node and benchmark it with IOR-3.3.0. The OrangeFS client and server are running on same machine. The OrangeFS version is 2.9.8. The kernel version is 3.10. I also tested with kernel 5.18.5, it has the same situation.
But I get extremely low performance when running on top of HDD, SSD, and NVMe respectively. The underlying file system is XFS.
(Here the I/O size is 4MB). During the test, I use orangeFS default configuration file (Command: pvfs2-genconfig --quiet --protocol tcp --tcpport 3334 --ioservers symbios-head --metaservers symbios-head --storage /mnt/nvme/jye20/OFS1-storage/data --metadata /mnt/nvme/jye20/OFS1-storage/meta --logfile /mnt/nvme/jye20/OFS1-storage/orangefs-server.log /home/jye20/symbios/ofs_install_scripts/pfs.conf)

OrangeFS's write and read performance on HDD:

IOR-3.3.0: MPI Coordinated Test of Parallel I/O
Began               : Tue Jun 21 11:54:50 2022
Command line        : /home/cc/common/ior-3.3.0/bin/ior -a POSIX --posix.odirect -t 4m -b 256m -s 1 -w -i 3 -F -k -m -e -d 0.1 -o /mnt/hdd/OFS1-mount/ior_data.dat
Machine             : Linux orangefs1.novalocal
TestID              : 0
StartTime           : Tue Jun 21 11:54:50 2022
Path                : /mnt/hdd/OFS1-mount
FS                  : 558.6 GiB   Used FS: 2.2%   Inodes: 8796093022208.0 Mi   Used Inodes: 0.0%

Options: 
api                 : POSIX
apiVersion          : 
test filename       : /mnt/hdd/OFS1-mount/ior_data.dat
access              : file-per-process
type                : independent
segments            : 1
ordering in a file  : sequential
ordering inter file : no tasks offsets
nodes               : 1
tasks               : 1
clients per node    : 1
repetitions         : 3
xfersize            : 4 MiB
blocksize           : 256 MiB
aggregate filesize  : 256 MiB

Results: 

access    bw(MiB/s)  IOPS       Latency(s)  block(KiB) xfer(KiB)  open(s)    wr/rd(s)   close(s)   total(s)   iter
------    ---------  ----       ----------  ---------- ---------  --------   --------   --------   --------   ----
write     186.75     47.51      0.021047    262144     4096       0.023772   1.35       0.000007   1.37       0   
write     154.96     39.16      0.025534    262144     4096       0.017886   1.63       0.000005   1.65       1   
write     169.67     42.93      0.023295    262144     4096       0.017933   1.49       0.000007   1.51       2   
Max Write: 186.75 MiB/sec (195.82 MB/sec)

IOR-3.3.0: MPI Coordinated Test of Parallel I/O
Began               : Tue Jun 21 11:54:57 2022
Command line        : /home/cc/common/ior-3.3.0/bin/ior -a POSIX --posix.odirect -t 4m -b 256m -s 1 -r -i 3 -F -k -m -e -d 0.1 -o /mnt/hdd/OFS1-mount/ior_data.dat
Machine             : Linux orangefs1.novalocal
TestID              : 0
StartTime           : Tue Jun 21 11:54:57 2022
Path                : /mnt/hdd/OFS1-mount
FS                  : 558.6 GiB   Used FS: 2.4%   Inodes: 8796093022208.0 Mi   Used Inodes: 0.0%

Options: 
api                 : POSIX
apiVersion          : 
test filename       : /mnt/hdd/OFS1-mount/ior_data.dat
access              : file-per-process
type                : independent
segments            : 1
ordering in a file  : sequential
ordering inter file : no tasks offsets
nodes               : 1
tasks               : 1
clients per node    : 1
repetitions         : 3
xfersize            : 4 MiB
blocksize           : 256 MiB
aggregate filesize  : 256 MiB

Results: 

access    bw(MiB/s)  IOPS       Latency(s)  block(KiB) xfer(KiB)  open(s)    wr/rd(s)   close(s)   total(s)   iter
------    ---------  ----       ----------  ---------- ---------  --------   --------   --------   --------   ----
read      227.91     57.67      0.017340    262144     4096       0.000096   1.11       0.013397   1.12       0   
read      182.11     45.77      0.021849    262144     4096       0.000088   1.40       0.007381   1.41       1   
read      167.43     42.33      0.023625    262144     4096       0.000069   1.51       0.016946   1.53       2   

OrangeFS's write and read performance on SSD:

Summary of tests:
Operation   Max(MiB)   Min(MiB)  Mean(MiB)     StdDev   Max(OPs)   Min(OPs)  Mean(OPs)     StdDev    Mean(s) Stonewall(s) Stonewall(MiB) Test# #Tasks tPN reps fPP reord reordoff reordrand seed segcnt   blksiz    xsize aggs(MiB)   API RefNum
write         339.76     336.19     338.39       1.57      84.94      84.05      84.60       0.39    0.75655         NA            NA     0      1   1    3   1     0        1         0    0      1 268435456  4194304     256.0 POSIX      0
read          506.52     487.98     499.35       8.13     126.63     122.00     124.84       2.03    0.51280         NA            NA     0      1   1    3   1     0        1         0    0      1 268435456  4194304     256.0 POSIX      0

OrangeFS's write and read performance on NVMe:

Summary of tests:
Operation   Max(MiB)   Min(MiB)  Mean(MiB)     StdDev   Max(OPs)   Min(OPs)  Mean(OPs)     StdDev    Mean(s) Stonewall(s) Stonewall(MiB) Test# #Tasks tPN reps fPP reord reordoff reordrand seed segcnt   blksiz    xsize aggs(MiB)   API RefNum
write         735.09     727.54     731.41       3.08     183.77     181.89     182.85       0.77    0.35002         NA            NA     0      1   1    3   1     0        1         0    0      1 268435456  4194304     256.0 POSIX      0
read         1249.51    1187.85    1228.38      28.67     312.38     296.96     307.10       7.17    0.20852         NA            NA     0      1   1    3   1     0        1         0    0      1 268435456  4194304     256.0 POSIX      0

Here are the hardware information:

OS: CentOS7, Kernel version: 3.10
HDD: Seagate / Module: ST600MP0005 / Size: 600GB 
SSD: Intel/ Module: SSDSC2KG019T8R / Size: 1.92TB
NVMe: Intel / Module: P3700 / Size: 2TB / Interface: PCIe

Trying to figure out where I went wrong....

**Comment by dreynol** [comment]
Hello Jie Ye,

Thank you for your interest in OrangeFS. Do you mind if I ask what you are looking to use the file system to accomplish?  OFS isn’t designed for environments where a single client is accessing a file on a single server, especially when the client and server are on a shared node and file accesses are relatively small. While a setup like that is fine for getting started with the file system and testing the functionality, benchmarking OFS in those conditions wouldn’t be expected to provide a high level of performance compared to a file system that was designed for serial workloads. Understanding your intended use case with OFS will help us lead you down the correct path.

– David

**Reply by Jye-525** [reply]
Hello David,

My research paper needs to compare the performance of OrangeFS with others. The performance of OrangeFS acts as a baseline. 
Here I only show the results of OFS on a single node. I also tested the performance of OrangeFS2.9.7 on Centos7 Linux 3.10.0-862.el7.x86_64 x86_64 with 8 OFS clients, and different scale OFS servers (4, 8, 12, 16, 24) before. The performance of OrangeFS is also very low.  The read performance is lower than its write performance, especially on NVMe. I am not sure what's wrong with this.

By the way, the stripe size is 4MB in this test. I use FIO to test orangeFS.

[OFS_Baseline_Final - OFS_Final_results.pdf](https://github.com/waltligon/orangefs/files/8968987/OFS_Baseline_Final.-.OFS_Final_results.pdf)

**Reply by dreynol** [reply]
For the best performance, we would highly recommend using a newer kernel that includes the OrangeFS kernel module. In your original message, I see where you said that you tested with 5.18.5 and got the same results. Were you using the kernel module in that case? With the upstream kernel module, you should have definitely seen significantly improved performance with 5.18.5.
